source/server/hot_restart_impl.cc

```diff
@@ -248,20 +248,22 @@ void HotRestartImpl::getParentStats(GetParentStatsInfo& info) {
   // could also potentially use connection oriented sockets and accept connections from our child,
   // and connect to our parent, but again, this becomes complicated.
   //
-  // Instead, we guard this condition with a lock. However, to avoid deadlock, we must try_lock()
+  // Instead, we guard this condition with a lock. However, to avoid deadlock, we must tryLock()
   // in this path, since this call runs in the same thread as the event loop that is receiving
-  // messages. If try_lock() fails it is sufficient to not return any parent stats.
-  std::unique_lock<Thread::BasicLockable> lock(init_lock_, std::defer_lock);
+  // messages. If tryLock() fails it is sufficient to not return any parent stats.
   memset(&info, 0, sizeof(info));
-  if (options_.restartEpoch() == 0 || parent_terminated_ || !lock.try_lock()) {
+  if (options_.restartEpoch() == 0 || parent_terminated_) {
     return;
   }
 
-  RpcBase rpc(RpcMessageType::GetStatsRequest);
-  sendMessage(parent_address_, rpc);
-  RpcGetStatsReply* reply = receiveTypedRpc<RpcGetStatsReply, RpcMessageType::GetStatsReply>();
-  info.memory_allocated_ = reply->memory_allocated_;
-  info.num_connections_ = reply->num_connections_;
+  Thread::TryLockGuard lock(init_lock_);
```

Comment:
Any reason for changing the control flow of this function, instead of just relacing replacing "!lock.try_lock()" with "!lock.isLocked()"?

Reply:
I think this flow is easier to reason about from an annotations analysis perspective.  Instantiating the TryLockGuard conceptually takes the lock -- though it may not actually acquire it due to try semantics.  I didn't want to attempt to acquire the lock if restartEpoch() was 0 or parent_terminated_, because it was not being acquired in the current code.  So this code has the same execution flow in terms of when the lock is taken, but binds it more clearly to the scope of the object.

Maybe a better model is to throw if the lock failed, and put a catch here?

Also if you prefer I can use early-exit instead of nesting the code to run when the lock succeeds:
```
if (!lock.isLocked()) {
  return;
}
```
I just can't put it in the same `||` expression as it was due to the timing of the instantiation of the scope.

Reply:
Any more thoughts @ggreenway? Happy to discuss further.

Reply:
Got it. I think this is fine.